webgme/webgme-core.d.ts

```diff
@@ -0,0 +1,816 @@
+
+/**
+Describe plugins
+*/
+declare namespace Core {
+
+    export type ISO8601 = string;
+    export type ErrorStr = string;
+    export type MetadataHash = string;
+    export type MetadataHashArray = string[];
+    export type ArtifactHash = string;
+    export type Name = string;
+
+    export type Metadata = {}
+    export type Node = {}
+    export type DataObject = {}
+    export type Buffer = GLbyte[];
+    export type Payload = string | Buffer | Buffer[];
+    export type Content = DataObject | Buffer | Buffer[];
+    export type ContentString = string;
+    export type Primitive = string | number;
+    export type OutAttr = DataObject | Primitive | undefined | null;
+    export type InAttr = DataObject | Primitive | null;
+    export type OutPath = string | undefined | null;
+
+    export type VoidFn = () => void;
+
+    export type MetadataHashCallback = (err: Error, result: MetadataHash) => void;
+    export type MetadataHashArrayCallback = (err: Error, result: MetadataHashArray) => void;
+    export type MetadataCallback = (err: Error, result: Metadata) => void;
+    export type ArtifactCallback = (err: Error, result: Artifact) => void;
+    export type ObjectCallback = (err: Error, result: DataObject) => void;
+    export type ObjectArrayCallback = (err: Error, result: DataObject[]) => void;
+    export type JSONCallback = (err: Error, result: JSON) => void;
+
+    export interface ResultCallback {
+        (err: Error | null, result: Result): void;
+    }
+
+    export interface Message {
+        msg: string;
+    }
+
+    export interface Artifact {
+        name: Name;
+        blobClient: BlobClient;
+        descriptor: Blob.BlobMetadata;
+
+        constructor(name: Name, blobClient: BlobClient, descriptor: Blob.BlobMetadata): void;
+
+        /** Adds content to the artifact as a file. */
+        addFile: {
+            (name: Name, content: Blob.ObjectBlob, callback: MetadataHashCallback): void;
+            (name: Name, content: Blob.ObjectBlob): Promise<MetadataHash>;
+        }
+        /** Adds files as soft-link. */
+        addFileAsSoftLink: {
+            (name: Name, content: Blob.ObjectBlob, callback: MetadataHashCallback): void;
+            (name: Name, content: Blob.ObjectBlob): Promise<MetadataHash>;
+        }
+        /** Adds multiple files. */
+        addFiles: {
+            (files: { [name: string]: Blob.ObjectBlob }, callback: MetadataHashArrayCallback): void;
+            (files: { [name: string]: Blob.ObjectBlob }): Promise<MetadataHashArray> | Promise<string>;
+        }
+        /** Adds multiple files as soft-links. */
+        addFilesAsSoftLinks: {
+            (files: { [name: string]: Blob.ObjectBlob }, callback: MetadataHashArrayCallback): void;
+            (files: { [name: string]: Blob.ObjectBlob }): Promise<MetadataHashArray>;
+        }
+        /** Adds a metadataHash to the artifact using the given file path. */
+        addMetadataHash: {
```

Comment:
This is an unusual way of writing a method with several overloads. Why not just `addMetadataHash(...): void; addMetadataHash(...): Promise<MetadataHash>`?